### PR TITLE
AdaptiveGrid SubtractBox some of edges were skipped - bugfix 

### PR DESCRIPTION
**Pull request by mwojcik-sagiton**
BACKGROUND:
During my work on routing function I noticed that some of edges are not removed after use of `AdaptiveGrid.SubtractBox(BBox3 box)`.
image

My first thought was that something is wrong with my configuration, but I could reproduce this bug in other routing functions, so the root cause was in `AdaptiveGrid` object.

DESCRIPTION:
I noticed that some of edge vertices are skipped because they are slightly misaligned and they are outside of `BBox3`. Therefore  method `OrientationTolerance()` returned ` PointOrientation.Low` or `PointOrientation.Hi`. This function was designed to return those values also for points that are on the edge of `BBox3`. My solution introduces another `PointOrientation` enum value `OnEdge` and checks if vertex is on the edge within grid tolerance. Due to additional enum I had to completely change logic of selection edges for deletion. I order to keep function working as before I had to add few more assertions in `AdaptiveGridSubtractBoxSmallDifference` test. Despite that I added two more tests, which are reproducing a bug:
- `AdaptiveGridSubtractManyBoxesWithSmallDifference` I tried to mock situation where bug occurred, but it was working without any changes 
- `AdaptiveGridSubtractMisalignedPolygon` this is part of grid from routing function that broke.

While I was preparing unit tests that will reproduce a bug I dicovered why some of edge vertexes are misaligned with the other vertexes. Root cause is small misaligned in bounding polygon used to add points to grid in method `AdpativeGrid.AddFromPolygon(boundary, points)`. I haven't fixed root cause in this PR, only side effects of it for `SubtractBox` function and I'm not sure where those side effects might appear. On the other hand I think that this bugfix is still worth merging because vertices might be misaligned from any other reasons and `SubtractBox` function should handle that too.

UPDATE: 
I reverted change with additional enum `OnEdge`, this change had too many side effects on previous logic. New solution fixes `Line.Intersects(BBox3 ....)` method which wasn't returning correct results when line vertices were misaligned with bounding box . 

TESTING:
Run `AdaptiveGridTests` and `AdaptiveGraphRoutingTests` unit tests. 
  
FUTURE WORK:
Remove misalignment root course, which is handling misaligned polygons in `AdapativeGrid.AddFromPolygon(boundary, points)` function.

REQUIRED:
- [X] All changes are up to date in `CHANGELOG.md`.

COMMENTS:
- Any other notes.

---
This change is [Reviewable](https://reviewable.io/reviews/hypar-io/elements/814)

### CODE REVIEW

**Review by wynged — CHANGES_REQUESTED** [review]

File: Elements/src/Spatial/AdaptiveGrid/AdaptiveGrid.cs [hunk 1]
```diff
@@ -183,70 +184,7 @@ public HashSet<Edge> AddFromPolygon(Polygon boundingPolygon, IEnumerable<Vector3
...
-                            edgesToDelete.Add(edge);
-                        }
-                    }
-                    if (intersections.Count == 2)
-                    {
-                        edgesToDelete.Add(edge);
-                    }
-                }
-            }
+            var edgesToDelete = GetEdges().Where(x => ShouldDeleteEdge(x, box));
```
[the remark this answers is not in the document]
Comment (mwojcik-sagiton) [reply Elements/src/Spatial/AdaptiveGrid/AdaptiveGrid.cs]: I reverted previous changes and extended unit tests for bounding box.

**Review by wynged — CHANGES_REQUESTED** [review]
Couple of notes about tests, mostly looks ok just going to ask for Dmytro review as well.

> Reviewed 1 of 3 files at r1, 4 of 4 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/hypar-io/elements/814#-:-N4Uxu4FBhoI0vs5sHYT:b-w75la7)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @mwojcik-sagiton)

___
*[`Elements/src/Geometry/Line.cs` line 425 at r3](https://reviewable.io/reviews/hypar-io/elements/814#-N4UxBWA0O0kJF_coLZd:-N4UxBWA0O0kJF_coLZe:bhuqyot) ([raw file](https://github.com/hypar-io/elements/blob/044f2388b2c632844545bed8a3998fc32b1a0b5b/Elements/src/Geometry/Line.cs#L425)):*
> ```Smalltalk
> 
>             // Check if found parameters are within normalized line range.
>             if (infinite || (tMin > -2 * Vector3.EPSILON && tMin < 1 + 2 * Vector3.EPSILON))
> ```

add a comment above about why we need this 2x factor.
___
*[`Elements/test/AdaptiveGridTests.cs` line 248 at r3](https://reviewable.io/reviews/hypar-io/elements/814#-N4UxLQ060pW4Nk-hGp6:-N4UxLQ060pW4Nk-hGp7:b-swc0h6) ([raw file](https://github.com/hypar-io/elements/blob/044f2388b2c632844545bed8a3998fc32b1a0b5b/Elements/test/AdaptiveGridTests.cs#L248)):*
> ```Smalltalk
>                 new Vector3(-29.8, 41.540211303259035, 8.37),
>                 
>                 //1st BBox vertices
> ```

do these boxes have different purposes?  could we comment the role that each of them has in testing?  for example "misaligned along x-axis"  "touches edge on one vertex"  "outside bounding box by tolerance exactly along entire edge"  
___
*[`Elements/test/LineTests.cs` line 401 at r3](https://reviewable.io/reviews/hypar-io/elements/814#-N4Uxftc7iBtYgWepJhl:-N4Uxftd0mTUHMAUULO_:b-goycni) ([raw file](https://github.com/hypar-io/elements/blob/044f2388b2c632844545bed8a3998fc32b1a0b5b/Elements/test/LineTests.cs#L401)):*
> ```Smalltalk
> 
>             //5. Line touches two sides of box and is misaligned slightly
>             var newBox = new BBox3(new Vector3(-30.41029, 19.60979, 8.37), new Vector3(-29.58971, 20.39021, 8.37));
> ```

can we create this mis-aligned box using the Vector3 tolerance itself?  so either add/subtract 1/2 tolerance to a specific vertex.

**Review by mwojcik-sagiton — COMMENTED** [review]
thanks :+1: 

> *[Reviewable](https://reviewable.io/reviews/hypar-io/elements/814#-:-N4YccL_BCtpTVXC13Vz:b-brn19m)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @mwojcik-sagiton)

**Review by DmytroMuravskyi — COMMENTED** [review]
> Reviewed 1 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/hypar-io/elements/814)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @mwojcik-sagiton and @wynged)

___
*[`Elements/src/Geometry/Line.cs` line 425 at r3](https://reviewable.io/reviews/hypar-io/elements/814#-N4UxBWA0O0kJF_coLZd:-N4fyrIXByxsnwZTMjF5:b-k29odi) ([raw file](https://github.com/hypar-io/elements/blob/044f2388b2c632844545bed8a3998fc32b1a0b5b/Elements/src/Geometry/Line.cs#L425)):*
<details><summary><i>Previously, wynged (Eric Wassail) wrote…</i></summary><blockquote>

add a comment above about why we need this 2x factor.
</blockquote></details>

Yes this is problematic place indeed. I don't think doubling the tolerance would help in general. Problem appears because checks before this function are done on line coordinates and here it's on [0;1] domain. If line has 0.75 length it will help but if length is 0.35, for example you would already need 3x tolerance to fit. What need to be done here is to multiply tMin and tMax by line length. 
This would actually sync these tolerance changes with ones performed in substructBox.
I should have done it in a very beginning, sorry.



_Code snippet:_
```Smalltalk
            var length = d.Length();
            var dMin = tMin * length;
            var dMax = tMax * length;
            // Check if found parameters are within normalized line range.
            if (infinite || (dMin > -Vector3.EPSILON && dMin < length + Vector3.EPSILON))
            {
                results.Add(Start + d * tMin);
            }

            if (Math.Abs(dMax - dMin) > Vector3.EPSILON &&
                (infinite || (dMax > -Vector3.EPSILON && dMax < length + Vector3.EPSILON)))
            {
                results.Add(Start + d * tMax);
            }
```

**Review by wynged — COMMENTED** [review]
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/hypar-io/elements/814)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @mwojcik-sagiton and @wynged)

**Review by mwojcik-sagiton — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hypar-io/elements/814)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @mwojcik-sagiton and @wynged)

___
*[`Elements/src/Geometry/Line.cs` line 425 at r3](https://reviewable.io/reviews/hypar-io/elements/814#-N4UxBWA0O0kJF_coLZd:-N50bjZo-0OYd4XhXhJx:b-egf729) ([raw file](https://github.com/hypar-io/elements/blob/044f2388b2c632844545bed8a3998fc32b1a0b5b/Elements/src/Geometry/Line.cs#L425)):*
<details><summary><i>Previously, DmytroMuravskyi wrote…</i></summary><blockquote>

Yes this is problematic place indeed. I don't think doubling the tolerance would help in general. Problem appears because checks before this function are done on line coordinates and here it's on [0;1] domain. If line has 0.75 length it will help but if length is 0.35, for example you would already need 3x tolerance to fit. What need to be done here is to multiply tMin and tMax by line length. 
This would actually sync these tolerance changes with ones performed in substructBox.
I should have done it in a very beginning, sorry.
</blockquote></details>

thank you @dmytromuravskyi your code works with my unit tests

**Review by wynged — COMMENTED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/hypar-io/elements/814)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @mwojcik-sagiton)

___
*[`Elements/test/LineTests.cs` line 401 at r3](https://reviewable.io/reviews/hypar-io/elements/814#-N4Uxftc7iBtYgWepJhl:-N50iMb15sB3BoSc4pL2:b-f3hm75) ([raw file](https://github.com/hypar-io/elements/blob/044f2388b2c632844545bed8a3998fc32b1a0b5b/Elements/test/LineTests.cs#L401)):*
<details><summary><i>Previously, wynged (Eric Wassail) wrote…</i></summary><blockquote>

can we create this mis-aligned box using the Vector3 tolerance itself?  so either add/subtract 1/2 tolerance to a specific vertex.
</blockquote></details>

we don't have to do this one.

**Review by wynged — APPROVED** [review]
> Reviewed 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/hypar-io/elements/814)* status: :shipit: complete! 1 of 1 approvals obtained

**Review by wynged — APPROVED** [review]
> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/hypar-io/elements/814)* status: :shipit: complete! 1 of 1 approvals obtained

**Review by wynged — APPROVED** [review]
> Reviewed 3 of 3 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/hypar-io/elements/814)* status: :shipit: complete! 1 of 1 approvals obtained

**Pull request merged**